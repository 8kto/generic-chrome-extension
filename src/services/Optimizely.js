// eslint-disable-next-line no-unused-vars
class Optimizely {
  /**
   * @param {string} cookies Document cookies
   */
  constructor(cookies) {
    this.cookies = cookies
    /** @type {Record<string, Experiment>} */
    this.experiments = {}
  }

  /**
   * @throws {Error}
   * @return {boolean}
   */
  isFeatureFlagsValid() {
    const experimentsCookie = this.cookies
      .split(';')
      .filter(i => i.match(/feature-flag-cookie/))

    if (!experimentsCookie.length) {
      throw new Error('No feature-flag-cookie found')
    }

    if (experimentsCookie.length !== 1) {
      throw new Error(
        'Ambiguous feature-flag-cookie found: remove multiple values and reload tab'
      )
    }

    try {
      const featureFlags = this.extractExperiments()

      return !!(
        typeof featureFlags === 'object' &&
        !Array.isArray(featureFlags) &&
        featureFlags &&
        Object.keys(featureFlags).length
      )
    } catch (err) {
      return false
    }
  }

  extractExperiments() {
    const experimentsCookie = this.cookies
      .split(';')
      .filter(i => i.match(/feature-flag-cookie/))

    if (experimentsCookie.length) {
      const json = experimentsCookie.shift().replace(/^[^{]+/, '')

      try {
        this.experiments = JSON.parse(json)
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error(error)
      }
    }

    return this.experiments
  }

  /**
   * @return {Record<string, Experiment>}
   */
  getExperiments() {
    return this.experiments
  }

  /**
   *
   * @param {string} experimentName
   * @param {boolean} status Enabled/Disabled
   */
  setExperimentStatus(experimentName, status) {
    this.experiments[experimentName].e = Boolean(status)

    return this
  }

  /**
   * @param {string} experimentName
   * @param {string} variableName
   * @param {any} value
   */
  setExperimentVariable(experimentName, variableName, value) {
    const experiment = this.experiments[experimentName]

    if (experiment) {
      experiment.v[variableName] = value
    }

    return this
  }

  /**
   * @param name
   * @param variant
   * @return {Record<string, Experiment>}
   */
  addNewExperiment(name, variant) {
    /** @type {Experiment} */
    this.experiments[name] = {
      e: true,
      v: { v_name: variant },
    }

    return this.experiments
  }

  isAvailable() {
    return Boolean(Object.keys(this.experiments || {}).length)
  }
}

/**
 * @typedef Variant
 * @property {string} v_name
 */

/**
 * @typedef Experiment
 * @property {boolean} e
 * @property {Variant} v
 */

// TODO review service, make static methods?
