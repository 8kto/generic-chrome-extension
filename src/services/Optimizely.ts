import type { ExperimentsList } from 'types'

import ChromeApi from './ChromeApi'

export default class Optimizely {
  // Document cookies
  private cookies: string
  private experiments: ExperimentsList = {}

  constructor(cookies: string) {
    this.cookies = cookies
  }

  /**
   * @throws {Error}
   */
  checkFeatureFlags(): void {
    const experimentsCookie = this.cookies
      .split(';')
      .filter(i => i.match(/feature-flag-cookie/))

    if (!experimentsCookie.length) {
      throw new Error('No feature-flag-cookie found')
    }

    if (experimentsCookie.length !== 1) {
      throw new Error(
        'Ambiguous feature-flag-cookie found: remove multiple values and reload tab'
      )
    }

    const featureFlags = this.extractExperiments()
    const isValid =
      typeof featureFlags === 'object' &&
      !Array.isArray(featureFlags) &&
      featureFlags &&
      Object.keys(featureFlags).length

    if (!isValid) {
      throw new Error(`Feature flags JSON is invalid`)
    }
  }

  extractExperiments(): ExperimentsList {
    const experimentsCookie = this.cookies
      .split(';')
      .filter(i => i.match(/feature-flag-cookie/))

    if (experimentsCookie.length) {
      const json = experimentsCookie.shift().replace(/^[^{]+/, '')

      try {
        this.experiments = JSON.parse(json)
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error(error)
      }
    }

    return this.experiments
  }

  getExperiments(): ExperimentsList {
    return this.experiments
  }

  setExperimentStatus(
    experimentName: string,
    status: boolean | string
  ): Optimizely {
    const normalizedBool =
      typeof status === 'boolean'
        ? status
        : {
            'true': true,
            'false': false,
          }[status]

    if (this.experiments[experimentName]) {
      this.experiments[experimentName].e = normalizedBool
    }

    return this
  }

  setExperimentVariable(
    experimentName: string,
    variableName: string,
    value: unknown
  ): Optimizely {
    const experiment = this.experiments[experimentName]

    if (experiment) {
      experiment.v[variableName] = value
    }

    return this
  }

  addNewExperiment(name: string, variant: string): ExperimentsList {
    /** @type {Experiment} */
    this.experiments[name] = {
      e: true,
      v: { v_name: variant },
    }

    return this.experiments
  }

  isAvailable(): boolean {
    return Boolean(Object.keys(this.experiments || {}).length)
  }

  static setFeatureFlagCookie(tabId: number, payload: string) {
    ChromeApi.executeScript({
      args: [payload],
      target: { tabId },
      // NB: it is not the usual closure, it doesn't capture any context
      function(payload: string) {
        document.cookie = `feature-flag-cookie=${payload}; path=/;`
      },
    })
  }

  static resetFeatureFlagCookie(
    tabId: number,
    callback: chrome.scripting.ScriptInjectionResultsHandler
  ): void {
    ChromeApi.executeScript(
      {
        target: { tabId },
        // NB: it is not the usual closure, it doesn't capture any context
        function() {
          ;[
            `feature-flag-cookie`,
            `feature-flag-user-token`,
            `feature-flag-targeting`,
          ].forEach(cookieName => {
            document.cookie =
              cookieName + '=;expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/;'
          })
        },
      },
      callback
    )
  }
}

// TODO review service, make static methods?
