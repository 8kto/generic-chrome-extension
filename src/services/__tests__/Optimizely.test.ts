import Optimizely from './../Optimizely'

const VALID_FF_STRING = JSON.stringify({
  'MOS-6502': {
    'e': true,
    'v': {
      'v_name': 'v3',
    },
  },
  'Z80': {
    'e': true,
    'v': {
      'v_name': 'variation_2',
    },
  },
})
const VALID_COOKIE = `my-cookie=42; feature-flag-cookie=${VALID_FF_STRING}`

describe('optimizely service', () => {
  // Just to disable console spamming during tests
  jest.spyOn(console, 'error').mockImplementation(() => undefined)

  describe('checkFeatureFlags', () => {
    let optimizelyService: Optimizely

    it('does not throw for valid cookies', () => {
      optimizelyService = new Optimizely(VALID_COOKIE)
      expect(() => optimizelyService.checkFeatureFlags()).not.toThrow()
    })

    it.each(['', '{}', 'null', 'undefined', '[]'])(
      'throws error for empty and falsy FF cookies',
      input => {
        optimizelyService = new Optimizely(
          `var=1;second=2;feature-flag-cookie=${input}`
        )
        expect(() => optimizelyService.checkFeatureFlags()).toThrow(
          'Feature flags JSON is invalid'
        )
      }
    )

    it('throws when no FF cookies', () => {
      optimizelyService = new Optimizely('var=1;second=2')
      expect(() => optimizelyService.checkFeatureFlags()).toThrow(
        'No feature-flag-cookie found'
      )
    })

    it('throws when multiple FF cookies', () => {
      optimizelyService = new Optimizely(
        `my-cookie=42; feature-flag-cookie=${VALID_FF_STRING}; feature-flag-cookie={}`
      )
      expect(() => optimizelyService.checkFeatureFlags()).toThrow(
        'Ambiguous feature-flag-cookie found: remove multiple values and reload tab'
      )
    })
  })

  describe('extractExperiments', () => {
    it('returns valid experiments', () => {
      const optimizelyService = new Optimizely(VALID_COOKIE)
      expect(optimizelyService.extractExperiments()).toMatchSnapshot(
        {},
        'valid'
      )
    })

    it('handles NA feature flags cookie', () => {
      const optimizelyService = new Optimizely('my-cookie=42;var=val')
      expect(optimizelyService.extractExperiments()).toStrictEqual({})
    })

    it.each(['', null, undefined])('handles falsy cookies [%s]', input => {
      const optimizelyService = new Optimizely(input)
      expect(optimizelyService.extractExperiments()).toStrictEqual({})
    })

    it('handles corrupted JSON in feature flags cookie', () => {
      const optimizelyService = new Optimizely(
        'my-cookie=42;feature-flag-cookie={"flag":{'
      )
      expect(optimizelyService.extractExperiments()).toStrictEqual({})
    })

    it.each(['', 'null', 'false', 'undefined', '[]', '{}', '"string"'])(
      'handles invalid feature flags cookie [%s]',
      input => {
        const optimizelyService = new Optimizely(
          'my-cookie=42;feature-flag-cookie=' + input
        )
        expect(optimizelyService.extractExperiments()).toStrictEqual({})
      }
    )
  })

  describe('setExperimentStatus', () => {
    let optimizelyService: Optimizely

    beforeEach(() => {
      optimizelyService = new Optimizely(VALID_COOKIE)
      optimizelyService.extractExperiments()
    })

    it('updates the experiment with bools', () => {
      optimizelyService.setExperimentStatus('MOS-6502', false)
      expect(optimizelyService.getExperiments()['MOS-6502'].e).toBe(false)

      optimizelyService.setExperimentStatus('MOS-6502', true)
      expect(optimizelyService.getExperiments()['MOS-6502'].e).toBe(true)
    })

    it('updates the experiment with strings', () => {
      optimizelyService.setExperimentStatus('MOS-6502', 'false')
      expect(optimizelyService.getExperiments()['MOS-6502'].e).toBe(false)

      optimizelyService.setExperimentStatus('MOS-6502', 'true')
      expect(optimizelyService.getExperiments()['MOS-6502'].e).toBe(true)
    })

    it('handles invalid values', () => {
      optimizelyService.setExperimentStatus('MOS-6502', undefined)
      expect(optimizelyService.getExperiments()['MOS-6502'].e).toBeUndefined()

      optimizelyService.setExperimentStatus('MOS-6502', null)
      expect(optimizelyService.getExperiments()['MOS-6502'].e).toBeUndefined()
    })

    it('does not throw on unknown experiment and corrupt data', () => {
      expect(() =>
        optimizelyService.setExperimentStatus('XXX', true)
      ).not.toThrow()
      expect(optimizelyService.getExperiments()).toMatchSnapshot({}, 'valid')
    })

    it('returns service instance', () => {
      const instance = optimizelyService.setExperimentStatus('MOS-6502', false)
      expect(instance).toBe(optimizelyService)
    })
  })

  // todo add return this cases

  describe('setExperimentVariable', () => {
    let optimizelyService: Optimizely

    beforeEach(() => {
      optimizelyService = new Optimizely(VALID_COOKIE)
      optimizelyService.extractExperiments()
    })

    const values = [
      'string',
      42,
      0xdeadbeaf,
      ['arr'],
      { key: 'val' },
      null,
      undefined,
      0,
      false,
      '',
    ]

    it.each(values)('updates value [%s]', input => {
      const instance = optimizelyService.setExperimentVariable(
        'Z80',
        'v_name',
        input
      )
      expect(optimizelyService.getExperiments()['Z80'].v['v_name']).toBe(input)
      expect(instance).toBe(optimizelyService)
    })

    it.each(values)('adds value [%s]', input => {
      const instance = optimizelyService.setExperimentVariable(
        'Z80',
        'new_var',
        input
      )
      expect(optimizelyService.getExperiments()['Z80'].v['new_var']).toBe(input)
      expect(instance).toBe(optimizelyService)
    })

    it('does not throw on unknown experiment', () => {
      expect(() =>
        optimizelyService.setExperimentVariable('XXX', 'test', 'fail')
      ).not.toThrow()
    })
  })
})
