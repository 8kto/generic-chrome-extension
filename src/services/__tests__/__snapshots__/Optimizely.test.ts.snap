// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optimizely service extractExperiments returns valid experiments: valid 1`] = `
Object {
  "MOS-6502": Object {
    "e": true,
    "v": Object {
      "v_name": "v3",
    },
  },
  "Z80": Object {
    "e": true,
    "v": Object {
      "v_name": "variation_2",
    },
  },
}
`;

exports[`optimizely service setExperimentStatus does not throw on unknown experiment and corrupt data: valid 1`] = `
Object {
  "MOS-6502": Object {
    "e": true,
    "v": Object {
      "v_name": "v3",
    },
  },
  "Z80": Object {
    "e": true,
    "v": Object {
      "v_name": "variation_2",
    },
  },
}
`;
